ggml-cuda.cu

```diff
@@ -225,6 +225,48 @@ static __global__ void dequantize_block_q8_0(const void * vx, float * y) {
     }
 }
 
+template <int block_size> static __global__ void dequantize_mul_mat_q4_0(const void * vx, const float * y, float * dst, const int ncols) {
+    const block_q4_0 * x = (const block_q4_0 *) vx;
+
+    const int row = blockIdx.x;
+    const int tid = threadIdx.x;
+
+    __shared__ float tmp[block_size]; // separate sum for each thread
+    tmp[tid] = 0;
+
+    for (int i = 0; i < ncols/block_size; i += 2) {
+        const int col = i*block_size + 2*tid;
+
+        // dequantize
+        const float d = x[(row*ncols + col)/QK4_0].d;
+
+        const uint8_t * pp = x[(row*ncols + col)/QK4_0].qs;
+
+        const uint8_t vui = pp[((row*ncols + col)%QK4_0)/2];
+
+        const int8_t vi0 = vui & 0xF;
+        const int8_t vi1 = vui >> 4;
+
+        const float v0 = (vi0 - 8)*d;
+        const float v1 = (vi1 - 8)*d;
+
+        // matrix multiplication
+        tmp[tid] += v0 * y[col + 0];
+        tmp[tid] += v1 * y[col + 1];
+    }
+
+    // sum up partial sums and write back result
+    for (int s=block_size/2; s>0; s>>=1) {
+        if (tid < s) {
+            tmp[tid] += tmp[tid + s];
+        }
+        __syncthreads();
+    }
```

Comment:
You don't need shared memory to sum results over a single warp; take a look at `warpReduceSum()` [here](https://github.com/NVIDIA/FasterTransformer/blob/c6e8f60ec40da218804a60e6aa986903e7fa8594/src/fastertransformer/kernels/reduce_kernel_utils.cuh#L74). If you want to use more warps in the future, you can do intra-warp sums first, then aggregates the partial sums between warps via shared memory (`blockReduceSum()` in the same file).

I can't say for sure that this will make a significant difference here, but I saw a lot of memory-bound kernels where getting rid of shared memory (or reducing the number of shared memory transactions) resulted in speedups.

Reply:
I read an official NVIDIA PDF where they said the same thing regarding intra-warp synchronization but when I removed the `__syncthreads()` instructions I got incorrect results even with a block size of 32. In any case, the addition of the partial sums at the end is negligible fore overall performance.

Reply:
> [...] when I removed the `__syncthreads()` instructions I got incorrect results even with a block size of 32

Is this code available somewhere? If you say that the addition overhead is negligible, then I guess it doesn't really matter in the big picture. I'm just curious if I can understand what went wrong.

Reply:
Sorry, I misremembered. I get incorrect results if I remove `__syncthreads()`. I get correct results again if I then also define `tmp` as volatile but this gives me worse performance. And when I tried just removing the summation alltogether the performance difference was negligible so I just kept `__syncthreads()`.

Reply:
It looks like we're talking about different things. [This](https://github.com/dfyz/llama.cpp/commit/44b3b3c53e7789ad1a13445c8617ab3bf5ce0090) is what I had in mind: no `__syncthreads()`, no `__shared__`, no `volatile`, no arrays.

On my A100, it does look like getting rid of shared memory doesn't improve anything on its own: the average time of running the kernel with `./main -b 512 -t 12 -n 256 -f prompts/dan.txt -m models/13B/ggml-model-q4_0.bin --no-mmap -s 123456 --gpu_layers 40` is **72** microseconds. However, this might be beneficial if you want to run more warps per block. For example:
  * [this variant](https://github.com/dfyz/llama.cpp/commit/b5ce497c0b8c80f841fc3ff39b1ae187916f6087) without shared memory runs in **64** microseconds on the same A100
  * [this variant](https://github.com/dfyz/llama.cpp/commit/f63760fe9f949e2322b80bb3c981b4ef27e23bc8) with shared memory runs in **68** microseconds on the same A100

Bottom line: this might not be very useful on its own, but reducing usage of shared memory is always a good idea in the log run.

Reply:
I tested a version without shared memory on my hardware. It was 3.5% times faster. I'm not going to optimize block sizes right now because I would like to do that at the end.